# Environment setup

Launch the following commands to setup a project environment:
> conda create --name db_project -c conda-forge --file requirements.txt

> conda activate db_project

> pip install dotenv

> pip install kagglehub

Write your database access credentials to the file `.env`

# Data preprocessing

In [10]:
import kagglehub
import pandas as pd
import numpy as np
import re

Oscars dataset:
https://www.kaggle.com/datasets/johnpendenque/oscars-from-1928-to-2024

In [11]:
oscars_path = kagglehub.dataset_download("johnpendenque/oscars-from-1928-to-2024")

In [12]:
oscars = pd.read_csv(oscars_path + "/oscars.csv")
oscars.head(3)

year  edition                    award   nomination_actor  \
0  2024       97  Actor In A Leading Role       Adrien Brody   
1  2024       97  Actor In A Leading Role  Timothée Chalamet   
2  2024       97  Actor In A Leading Role     Colman Domingo   

  nomination_country nomination_character_name nomination_citation  \
0                NaN               László Tóth                 NaN   
1                NaN                 Bob Dylan                 NaN   
2                NaN                  Divine G                 NaN   

  nomination_producers nomination_description          film_title  is_winner  \
0                  NaN                    NaN       The Brutalist       True   
1                  NaN                    NaN  A Complete Unknown      False   
2                  NaN                    NaN           Sing Sing      False   

   acceptance_speech_text                              acceptance_speech_url  
0                     NaN  http://aaspeechesdb.oscars.org/ics-wpd/url.ash...  
1                     NaN                                                NaN  
2                     NaN                                                NaN

In [13]:
rows_number, cols_number = oscars.shape
print(f"Oscars dataset contains {rows_number} rows and {cols_number} columns")

Oscars dataset contains 11995 rows and 13 columns


In [14]:
print(f"The percentage of NaN values in the film_title column: {round(oscars.film_title.isna().mean() * 100, 1)}%")

The percentage of NaN values in the film_title column: 10.9%


<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
    
Some award categories are not associated with a specific film. For example, the *Honorary Award* or the *Jean Hersholt Humanitarian Award*. Around 11% of the rows in the dataset fall into this group (i.e., *film_title* equals NaN). We will remove these rows.
</div>

In [15]:
oscars.dropna(subset=['film_title'], inplace=True)

In [16]:
print("The percentage of NaN values in the columns of the dataset:")
oscars.isna().mean().sort_values(ascending=False) * 100

The percentage of NaN values in the columns of the dataset:


nomination_citation          100.000000
nomination_description       100.000000
acceptance_speech_text       100.000000
nomination_country            99.719311
nomination_producers          87.846183
acceptance_speech_url         83.327096
nomination_character_name     82.718937
nomination_actor              12.453219
year                           0.000000
edition                        0.000000
award                          0.000000
is_winner                      0.000000
film_title                     0.000000
dtype: float64

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">

The columns *nomination_citation*, *nomination_description*, *acceptance_speech_text*, and *nomination_country* contain around 100% NaN values. Besides, the *acceptance_speech_url* data is not relevant for our analysis. We will remove these columns as well.
</div>

In [17]:
oscars.drop(
    columns=[
        'nomination_citation',
        'nomination_description',
        'acceptance_speech_text',
        'nomination_country',
        'acceptance_speech_url'
    ],
    inplace=True
)

In [18]:
print("The percentage of NaN values in the columns of the dataset:")
oscars.isna().mean().sort_values(ascending=False) * 100

The percentage of NaN values in the columns of the dataset:


nomination_producers         87.846183
nomination_character_name    82.718937
nomination_actor             12.453219
year                          0.000000
award                         0.000000
edition                       0.000000
film_title                    0.000000
is_winner                     0.000000
dtype: float64

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">

We can also notice that when the people involved in a nomination are specified in a row, they are listed either in the *nomination_actor* column or in the *nomination_producers* column.
</div>

In [19]:
any(~oscars.nomination_actor.isna().values & ~oscars.nomination_producers.isna().values)

False

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">

Thus, we can combine these two columns into one and name it *nomination_people*.
</div>

In [20]:
oscars['nomination_people'] = oscars.nomination_actor.fillna(oscars.nomination_producers)

In [21]:
oscars.drop(
    columns=[
        'nomination_actor',
        'nomination_producers',
    ],
    inplace=True
)

In [22]:
print("The percentage of NaN values in the columns of the dataset:")
oscars.isna().mean().sort_values(ascending=False) * 100

The percentage of NaN values in the columns of the dataset:


nomination_character_name    82.718937
nomination_people             0.299401
year                          0.000000
award                         0.000000
edition                       0.000000
film_title                    0.000000
is_winner                     0.000000
dtype: float64

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
    
The *nomination_people* column contains highly inconsistent textual data: sometimes a single person, sometimes multiple people, sometimes people sharing one last name, and sometimes entries that are not people at all (countries, groups, studios, etc.).
</div>

In [23]:
oscars['_space_count'] = oscars.nomination_people.str.count(' ')
for n in range(0, 11):
    subset = oscars[oscars['_space_count'] == n]['nomination_people'].head(3)
    examples = subset.tolist()
    print(f"Examples of nomination_people values containing {n} spaces:")
    for example in examples:
        print("    ", example)

oscars.drop(columns=['_space_count'], inplace=True)

Examples of nomination_people values containing 0 spaces:
     Lebanon
     Poland
     Germany
Examples of nomination_people values containing 1 spaces:
     Adrien Brody
     Timothée Chalamet
     Colman Domingo
Examples of nomination_people values containing 2 spaces:
     Karla Sofía Gascón
     Marc Platt, Producer
     Sterling K. Brown
Examples of nomination_people values containing 3 spaces:
     Clément Ducol and Camille
     Victoria Warmerdam and Trent
     Screenplay by Peter Straughan
Examples of nomination_people values containing 4 spaces:
     Kelsey Mann and Mark Nielsen
     Adam Elliot and Liz Kearney
     Chris Sanders and Jeff Hermann
Examples of nomination_people values containing 5 spaces:
     Nicolas Keppens and Brecht Van Elslande
     Pascal Caucheteux and Jacques Audiard, Producers
     Ema Ryan Yamazaki and Eric Nyari
Examples of nomination_people values containing 6 spaces:
     Nick Park, Merlin Crossingham and Richard Beek
     Maria Carlota Bruno and R

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Because of this wide variety of formats, we need a robust parsing function that can handle all these scenarios. The function below normalizes the text, detects and expands shared-last-name patterns, filters out non-human entries, and finally returns a clean list of individual people (each represented as "First Last") for every row.

Rules applied:
- A valid person must consist of at least two words (e.g., "John Smith").
- Single words (e.g., "Russia", "Pixar") are ignored.
- Group indicators (team, company, studio...) are ignored.
- The pattern "Joel and Ethan Coen" should be expanded to: \["Joel Coen", "Ethan Coen"\].
</div>

In [24]:
def parse_people(text):
    if pd.isna(text) or not isinstance(text, str) or not text.strip():
        return []

    t = text.strip()

    # 1. Normalize connectors: "&" -> "and" -> ","
    t = re.sub(r'\s*&\s*', ' and ', t, flags=re.I)
    t = re.sub(r'\s+and\s+', ', ', t, flags=re.I)

    # 2. Split on commas into individual candidate segments
    parts = [p.strip() for p in t.split(',') if p.strip()]

    persons = []
    for part in parts:
        # 3. Ignore group-like entities
        if re.search(r'\b(group|team|band|studio|company|committee|ensemble|crew)\b', part, flags=re.I):
            continue

        # 4. Handle cases with shared last name (e.g. "Joel and Ethan Coen")
        m = re.match(r'(.+?)\s+(\w+)$', part)
        if m:
            body, lastname = m.group(1), m.group(2)
            subparts = re.split(r'\s+and\s+', body)
            if len(subparts) > 1:
                for sp in subparts:
                    full_name = (sp.strip() + " " + lastname).strip()

                    # 5. Validate: a person must have >= 2 words
                    if len(full_name.split()) >= 2:
                        persons.append(full_name)
                continue

        # 6. Regular case: acept only segments that contain >= 2 words
        if len(part.split()) >= 2:
            persons.append(part)

    return persons

In [25]:
oscars['people_list'] = oscars.nomination_people.apply(parse_people)
empty_people_rows = oscars[oscars['people_list'].apply(lambda x: len(x) == 0)]
print(f"The dataset contains only {len(empty_people_rows)} rows in which people_list is empty")

The dataset contains only 481 rows in which people_list is empty


<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
    
We applied the parsing function to all rows, saved the result into the *people_list* column, and then extracted the rows for which the list of people turned out to be empty. There were only 481 such rows, so we saved them and manually verified that they indeed do not contain any human names and our function works as expected.
</div>

In [26]:
# empty_people_rows.to_csv('empty_rows.csv', index=False)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
    
Now we can remove the *nomination_people* column.
</div>

In [27]:
oscars.drop(
    columns=[
        'nomination_people',
    ],
    inplace=True
)

In [28]:
oscars.dtypes

year                          int64
edition                       int64
award                        object
nomination_character_name    object
film_title                   object
is_winner                      bool
people_list                  object
dtype: object

In [29]:
oscars.head()

year  edition                    award nomination_character_name  \
0  2024       97  Actor In A Leading Role               László Tóth   
1  2024       97  Actor In A Leading Role                 Bob Dylan   
2  2024       97  Actor In A Leading Role                  Divine G   
3  2024       97  Actor In A Leading Role                  Lawrence   
4  2024       97  Actor In A Leading Role              Donald Trump   

           film_title  is_winner          people_list  
0       The Brutalist       True       [Adrien Brody]  
1  A Complete Unknown      False  [Timothée Chalamet]  
2           Sing Sing      False     [Colman Domingo]  
3            Conclave      False      [Ralph Fiennes]  
4      The Apprentice      False     [Sebastian Stan]

# Database creation

In [30]:
import os
import io
from sqlalchemy import create_engine, text, DDL
from dotenv import load_dotenv
from sqlalchemy.exc import IntegrityError

In [31]:
load_dotenv()

True

In [32]:
password = os.getenv("PASSWORD")
user = os.getenv("USERNAME_DB")
db = os.getenv("NAME_DB")
engine = create_engine(f"postgresql+psycopg2://{user}:{password}@postgresql-edu.in.centralelille.fr:5432/{db}")

In [33]:
# uncomment if needed
with engine.begin() as conn:
   conn.execute(text("""DROP SCHEMA films CASCADE;"""))

In [34]:
with engine.begin() as conn:
    conn.execute(text("""CREATE SCHEMA films;"""))
    conn.execute(text("""set search_path = 'films'"""))
   
    conn.execute(text("""
    CREATE TABLE IF NOT EXISTS Person (
        person_id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
        name TEXT NOT NULL UNIQUE
    ); """))

    conn.execute(text("""
    CREATE TABLE IF NOT EXISTS Film (
        film_id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
        title TEXT NOT NULL,
        CONSTRAINT unique_title UNIQUE (title)
    ); """))

    conn.execute(text("""
    CREATE TABLE IF NOT EXISTS Award (
        award_id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
        award_name TEXT NOT NULL UNIQUE
    ); """))

    conn.execute(text("""
    CREATE TABLE IF NOT EXISTS Edition (
        year INTEGER PRIMARY KEY,
        edition INTEGER NOT NULL UNIQUE
    ); """))

    conn.execute(text("""
    CREATE TABLE IF NOT EXISTS Nomination (
        nom_id INTEGER GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
        year INTEGER NOT NULL,
        film_id INTEGER NOT NULL,
        award_id INTEGER NOT NULL,
        is_winner BOOL NOT NULL,
        UNIQUE(year, film_id, award_id),
        FOREIGN KEY(year) REFERENCES Edition(year) ON DELETE CASCADE,
        FOREIGN KEY(film_id) REFERENCES Film(film_id) ON DELETE CASCADE,
        FOREIGN KEY(award_id) REFERENCES Award(award_id) ON DELETE CASCADE
    ); """))

    conn.execute(text("""
    CREATE TABLE IF NOT EXISTS Nomination_person (
        nomination_id INTEGER NOT NULL,
        person_id INTEGER NOT NULL,
        character TEXT,
        PRIMARY KEY (nomination_id, person_id),
        FOREIGN KEY (nomination_id) REFERENCES Nomination(nom_id) ON DELETE CASCADE,
        FOREIGN KEY (person_id) REFERENCES Person(person_id) ON DELETE CASCADE
    ); """))

# Table population

In [35]:
# We need a low-level DB connection because we will use PostgreSQL's COPY via psycopg2 cursor.
# SQLAlchemy's high-level Connection doesn't expose copy_expert, so we get the raw DB connection.
raw_conn = engine.raw_connection()
cursor = raw_conn.cursor()

# Make a working copy of oscars dataset so that we don't modify it
df = oscars.copy()

# Convert people_list to string
df["people_list"] = df["people_list"].apply(
    lambda x: ",".join(map(str, x)) if isinstance(x, list) else str(x)
)

# Convert the DataFrame to an in-memory CSV. COPY from STDIN expects a file-like object.
csv_buffer = io.StringIO()
df.to_csv(csv_buffer, index=False, header=False)
csv_buffer.seek(0)  # rewind to start so copy_expert reads from the beginning

# Create a temporary table that mirrors the columns we will COPY.
cursor.execute("""
    DROP TABLE IF EXISTS tmp_oscars;
    CREATE TEMP TABLE tmp_oscars (
        year INTEGER,
        edition INTEGER,
        award TEXT,
        nomination_character_name TEXT,
        film_title TEXT,
        is_winner BOOLEAN,
        people_list TEXT
    );
""")

# Use psycopg2's copy_expert to run a COPY FROM STDIN.
# Because we passed header=False above, COPY will read rows in the exact column order we exported.
cursor.copy_expert(
    """
    COPY tmp_oscars
    FROM STDIN WITH (FORMAT CSV);
    """,
    csv_buffer
)

# Commit the COPY - required because we're using raw_connection outside SQLAlchemy transactional context.
raw_conn.commit()

In [36]:
# Insert Editions (one row per year + edition). DISTINCT prevents duplicate inserts.
# ON CONFLICT (year) DO NOTHING uses the unique constraint on year to skip existing rows.
cursor.execute("""
    INSERT INTO Edition (edition, year)
    SELECT DISTINCT edition, year 
    FROM tmp_oscars
    ON CONFLICT (year) DO NOTHING;
""")

In [37]:
# Insert unique awards from tmp_oscars -> Award table
cursor.execute("""
    INSERT INTO Award (award_name)
    SELECT DISTINCT award FROM tmp_oscars
    ON CONFLICT (award_name) DO NOTHING;
""")

In [38]:
# Insert unique films into Film(title)
cursor.execute("""
    INSERT INTO Film (title)
    SELECT DISTINCT film_title FROM tmp_oscars;
""")

In [39]:
# Insert nominations by joining tmp data to the Film and Award rows we just inserted.
cursor.execute("""
    INSERT INTO Nomination (year, film_id, award_id, is_winner)
    SELECT 
        t.year,
        f.film_id,
        a.award_id,
        t.is_winner
    FROM tmp_oscars t
    JOIN Film f ON f.title = t.film_title
    JOIN Award a ON a.award_name = t.award
    ON CONFLICT (year, film_id, award_id) DO NOTHING;
""")

In [40]:
# Expand people_list into separate rows and insert unique Person names
cursor.execute("""
    WITH expanded AS (
        SELECT 
            nom.nom_id,
            unnest(string_to_array(t.people_list, ',')) AS person_name,
            t.nomination_character_name AS character
        FROM tmp_oscars t
        JOIN Film f ON f.title = t.film_title
        JOIN Award a ON a.award_name = t.award
        JOIN Nomination nom 
          ON nom.year = t.year 
         AND nom.film_id = f.film_id 
         AND nom.award_id = a.award_id
    )
    INSERT INTO Person (name)
    SELECT DISTINCT trim(person_name)
    FROM expanded
    ON CONFLICT (name) DO NOTHING;
""")

# - string_to_array(t.people_list, ',') splits the comma-separated string into text[]
# - unnest(...) expands the array into rows
# - trim(...) removes leading/trailing whitespace from names

In [41]:
# Now insert the association rows into Nomination_person
cursor.execute("""
    WITH expanded AS (
        SELECT 
            nom.nom_id,
            trim(unnest(string_to_array(t.people_list, ','))) AS person_name,
            t.nomination_character_name AS character
        FROM tmp_oscars t
        JOIN Film f ON f.title = t.film_title
        JOIN Award a ON a.award_name = t.award
        JOIN Nomination nom 
          ON nom.year = t.year 
         AND nom.film_id = f.film_id 
         AND nom.award_id = a.award_id
    )
    INSERT INTO Nomination_person (nomination_id, person_id, character)
    SELECT 
        e.nom_id,
        p.person_id,
        e.character
    FROM expanded e
    JOIN Person p ON p.name = trim(e.person_name)
    ON CONFLICT (nomination_id, person_id) DO NOTHING;
""")

raw_conn.commit()
cursor.close()
raw_conn.close()

# Queries (part 1)

In [42]:
def run_query(query):
    text_query = text(query)
    return pd.read_sql_query(text_query, engine)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Collect all the Award names
</div>

In [43]:
run_query("""
SELECT award_name
FROM award
;""")

award_name
0    Music (Original Song Score And Its Adaptation ...
1                                  Engineering Effects
2                                         Film Editing
3                            Actress In A Leading Role
4                                  Outstanding Picture
..                                                 ...
102                                      Art Direction
103                       Directing (Dramatic Picture)
104                                    Special Effects
105                            Writing (Title Writing)
106         Special Achievement Award (Visual Effects)

[107 rows x 1 columns]

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Get the titles of all movies that have won any type of Oscar.
</div>

In [44]:
run_query("""
SELECT title
FROM film
JOIN nomination as nom ON film.film_id = nom.film_id
WHERE nom.is_winner
;""")

title
0     The Brutalist
1       A Real Pain
2             Anora
3      Emilia Pérez
4              Flow
...             ...
2168          Wings
2169          Wings
2170        Sunrise
2171     7th Heaven
2172     Underworld

[2173 rows x 1 columns]

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Count the number of times a film was nominated.
</div>

In [45]:
run_query("""
SELECT film.title, COUNT(Nomination.nom_id) AS nom_total
FROM film
JOIN nomination ON film.film_id = nomination.film_id
GROUP BY film.title
ORDER BY nom_total desc
;""")

title  nom_total
0                                 A Star Is Born         25
1                                West Side Story         18
2                                        Titanic         16
3                                   Moulin Rouge         15
4                                      Cleopatra         14
...                                          ...        ...
5085                          Information Please          1
5086                                   Crip Camp          1
5087  Journey into Life: The World of the Unborn          1
5088                         There Goes My Heart          1
5089                                  Short Cuts          1

[5090 rows x 2 columns]

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
List all categories in which Titanic was nominated.
</div>

In [46]:
run_query("""
SELECT award.award_name
FROM award
JOIN nomination ON nomination.award_id = award.award_id
JOIN film ON nomination.film_id = film.film_id
WHERE film.title = 'Titanic'
;""")

award_name
0         Actress In A Leading Role
1      Actress In A Supporting Role
2                     Art Direction
3                      Best Picture
4                    Cinematography
5                    Costume Design
6                         Directing
7                      Film Editing
8                            Makeup
9   Music (Original Dramatic Score)
10            Music (Original Song)
11                            Sound
12            Sound Effects Editing
13                   Visual Effects
14  Art Direction (Black-And-White)
15   Writing (Story And Screenplay)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Give the names of the people who played the Joker in a movie.
</div>

In [47]:
run_query("""
SELECT person.name
FROM person
JOIN nomination_person ON person.person_id = nomination_person.person_id
WHERE nomination_person.character = 'Joker'
;""")

name
0  Heath Ledger

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Give the names of the people who played a character whose name contains "Bob".
</div>

In [48]:
run_query("""
SELECT person.name, nomination_person.character
FROM person
JOIN nomination_person ON person.person_id = nomination_person.person_id
WHERE nomination_person.character LIKE '%Bob %'
;""")

name         character
0        Bill Murray        Bob Harris
1         Bruce Dern  Captain Bob Hyde
2     George Clooney        Bob Barnes
3  Timothée Chalamet         Bob Dylan

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Actors and actresses who were nominated for both a leading and a supporting role.
</div>

In [49]:
run_query("""
WITH t AS (
    SELECT p.name, a.award_name
    FROM person AS p
    JOIN nomination_person AS np ON p.person_id = np.person_id
    JOIN nomination AS n ON np.nomination_id = n.nom_id
    JOIN award AS a ON n.award_id = a.award_id
)
SELECT t.name
FROM t
WHERE t.award_name = 'Actor In A Leading Role' OR t.award_name = 'Actress In A Leading Role'
INTERSECT
SELECT t.name
FROM t
WHERE t.award_name = 'Actor In A Supporting Role' OR t.award_name = 'Actress In A Supporting Role'
;""")

name
0         Nick Nolte
1        Paul Newman
2       Kate Winslet
3      Robert Duvall
4        Viola Davis
..               ...
125       Matt Damon
126  Richard Jenkins
127      Adam Driver
128    Nicole Kidman
129     Helen Mirren

[130 rows x 1 columns]

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
The film with the highest win-to-nomination ratio.
</div>

In [50]:
run_query("""
WITH film_stats AS (
    SELECT film.title, n.year, COUNT(n.nom_id) AS total_noms, SUM(n.is_winner::INT) AS total_wins
    FROM film
    JOIN nomination AS n ON film.film_id = n.film_id
    GROUP BY film.title, n.year
    HAVING COUNT(n.nom_id) >= 5
)
SELECT title, year, total_wins, total_noms, total_wins / total_noms AS conversion_rate
FROM film_stats
ORDER BY conversion_rate DESC, total_noms DESC LIMIT 1
;""")

title  year  total_wins  \
0  The Lord of the Rings: The Return of the King  2003          11   

   total_noms  conversion_rate  
0          11                1

# New data

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Once the Oscars database has been set up, someone approaches us and asks whether it’s possible to add the film data they’ve gathered so that the Oscars can be linked to detailed film information.
</div>

Movies dataset:
https://www.kaggle.com/datasets/asaniczka/tmdb-movies-dataset-2023-930k-movies

In [51]:
films_path = kagglehub.dataset_download("asaniczka/tmdb-movies-dataset-2023-930k-movies")

In [52]:
films = pd.read_csv(films_path + '/TMDB_movie_dataset_v11.csv')
films.head(3)

id            title  vote_average  vote_count    status release_date  \
0   27205        Inception         8.364       34495  Released   2010-07-15   
1  157336     Interstellar         8.417       32571  Released   2014-11-05   
2     155  The Dark Knight         8.512       30619  Released   2008-07-16   

      revenue  runtime  adult                     backdrop_path  ...  \
0   825532764      148  False  /8ZTVqvKDQ8emSGUEMjsS4yHAwrp.jpg  ...   
1   701729206      169  False  /pbrkL804c8yAv3zBZR4QPEafpAR.jpg  ...   
2  1004558444      152  False  /nMKdUUepR0i5zn0y1T4CsSB5chy.jpg  ...   

    original_title                                           overview  \
0        Inception  Cobb, a skilled thief who commits corporate es...   
1     Interstellar  The adventures of a group of explorers who mak...   
2  The Dark Knight  Batman raises the stakes in his war on crime. ...   

  popularity                       poster_path  \
0     83.952  /oYuLEt3zVCKq57qu2F8dT7NIa6f.jpg   
1    140.241  /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg   
2    130.643  /qJ2tW6WMUDux911r6m7haRef0WH.jpg   

                                             tagline  \
0               Your mind is the scene of the crime.   
1  Mankind was born on Earth. It was never meant ...   
2                  Welcome to a world without rules.   

                               genres  \
0  Action, Science Fiction, Adventure   
1   Adventure, Drama, Science Fiction   
2      Drama, Action, Crime, Thriller   

                                production_companies  \
0  Legendary Pictures, Syncopy, Warner Bros. Pict...   
1  Legendary Pictures, Syncopy, Lynda Obst Produc...   
2  DC Comics, Legendary Pictures, Syncopy, Isobel...   

                       production_countries  \
0  United Kingdom, United States of America   
1  United Kingdom, United States of America   
2  United Kingdom, United States of America   

                     spoken_languages  \
0  English, French, Japanese, Swahili   
1                             English   
2                   English, Mandarin   

                                            keywords  
0  rescue, mission, dream, airplane, paris, franc...  
1  rescue, future, spacecraft, race against time,...  
2  joker, sadism, chaos, secret identity, crime f...  

[3 rows x 24 columns]

# Data preprocessing

In [53]:
rows_number, cols_number = films.shape
print(f"Films dataset contains {rows_number} rows and {cols_number} columns")

Films dataset contains 1334510 rows and 24 columns


<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
    
Convert *release_date* column to datetime
</div>

In [54]:
films['release_date'] = pd.to_datetime(films['release_date'], errors='coerce')

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
At first, we need to drop all the adult movies and the movies without any name
</div>

In [55]:
films = films[~films['adult']]
films = films.dropna(subset=['title'])
films.shape

(1202292, 24)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
We assume that no two films share the same title and release year, so we will remove duplicates.
</div>

In [56]:
# Count the number of duplicate
films['low_title'] = films['title'].str.lower()
nb_duplicate = films.duplicated(subset=['low_title', 'release_date']).sum()

print(f"The total number of films that share the same title and release year: {nb_duplicate}")

films.drop_duplicates(subset=['low_title', 'release_date'], keep='first', inplace=True)
films.shape

The total number of films that share the same title and release year: 26165


(1176127, 25)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
We remove all films that have not yet been released, since they could not have won an Oscar or participated in a nomination.
</div>

In [57]:
films = films.dropna(subset=['release_date'])
films.shape

(942551, 25)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
The table contains many irrelevant columns. We will keep only the following columns: 
<ul> 
    <li> title</li> 
    <li> release_date</li> 
    <li> vote_average</li> 
    <li> vote_count</li> 
    <li> runtime</li>
    <li> revenue</li> 
    <li> budget</li> 
</ul> </div>

In [58]:
columns = [
    'title',
    'release_date',
    'vote_average',
    'vote_count',
    'runtime',
    'revenue',
    'budget'
]
films = films[columns]
films.head()

title release_date  vote_average  vote_count  runtime  \
0        Inception   2010-07-15         8.364       34495      148   
1     Interstellar   2014-11-05         8.417       32571      169   
2  The Dark Knight   2008-07-16         8.512       30619      152   
3           Avatar   2009-12-15         7.573       29815      162   
4     The Avengers   2012-04-25         7.710       29166      143   

      revenue     budget  
0   825532764  160000000  
1   701729206  165000000  
2  1004558444  185000000  
3  2923706026  237000000  
4  1518815515  220000000

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Replace 0 values in budget, revenue, or runtime with NaN. Also, set vote_average to NaN for films with vote_count equal to 0.
</div>

In [59]:
films[['budget', 'runtime', 'revenue']] = films[['budget', 'runtime', 'revenue']].replace(0, np.nan)
films.loc[films['vote_count'] == 0, 'vote_average'] = np.nan

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
To determine which film corresponds to an Oscar database when there are two films with the same title, we select the film whose release date is closest to the year of the first nomination.
</div>

In [60]:
films_oscar = run_query("""
SELECT f.film_id, lower(f.title) as title_low, min(n.year) as first_oscar_year
FROM film f
JOIN nomination n ON f.film_id = n.film_id
GROUP BY f.title, f.film_id
;""")
films_oscar.head()

film_id               title_low  first_oscar_year
0     1489  take a letter, darling              1942
1     4790  avengers: infinity war              2018
2      273        the virgin queen              1955
3     3936             hunger ward              2020
4     2574                  exodus              1960

In [61]:
films['title_low'] = films['title'].str.lower()
df_shared_title = pd.merge(
    films_oscar,
    films,
    on='title_low',
    how='inner',
    suffixes=('_bdd', '_tmdb')
)
df_shared_title.head()

film_id               title_low  first_oscar_year                   title  \
0     1489  take a letter, darling              1942  Take a Letter, Darling   
1     4790  avengers: infinity war              2018  Avengers: Infinity War   
2      273        the virgin queen              1955        The Virgin Queen   
3      273        the virgin queen              1955        The Virgin Queen   
4      273        the virgin queen              1955        The Virgin Queen   

  release_date  vote_average  vote_count  runtime       revenue       budget  
0   1942-05-06         6.600           9     92.0           NaN          NaN  
1   2018-04-25         8.255       27713    149.0  2.052415e+09  300000000.0  
2   1955-07-22         7.130          23     92.0           NaN          NaN  
3   1928-05-12           NaN           0     20.0           NaN          NaN  
4   1923-01-23           NaN           0      NaN           NaN          NaN

In [62]:
df_shared_title['diff_year'] = abs(df_shared_title['release_date'].dt.year - df_shared_title.first_oscar_year)
df_shared_title.sort_values(by=['film_id', 'diff_year'], ascending=True, inplace=True)
df_shared_title.drop_duplicates(subset=['film_id'], keep='first', inplace=True)

In [63]:
final_data = df_shared_title.drop(columns=['title_low', 'first_oscar_year', 'diff_year'])

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
As we can see, out of the 5090 films already in the database, 357 had no additional information in the movie dataset.
</div>

In [64]:
final_data.shape

(4733, 8)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
The final_data dataset contains only films that are already in the Oscars database.
</div>

# Adding new data to the dataset

In [65]:
with engine.begin() as conn:
    conn.execute(text("""
    ALTER TABLE Film
        ADD COLUMN release_date DATE,
        ADD COLUMN vote_average FLOAT,
        ADD COLUMN vote_count INTEGER,
        ADD COLUMN runtime INTEGER,
        ADD COLUMN revenue NUMERIC,
        ADD COLUMN budget NUMERIC;
    """))

    conn.execute(text("""
    ALTER TABLE Film
        DROP CONSTRAINT unique_title;
    """))

    conn.execute(text("""
    ALTER TABLE Film
        ADD CONSTRAINT uq_film_title_release UNIQUE (title, release_date);
    """))

In [66]:
updates_cols = ['release_date', 'vote_average', 'vote_count', 'runtime', 'revenue', 'budget']
with engine.begin() as conn:
    for index, row in final_data.iterrows():
        set_clauses = []
        params = {}
        for col in updates_cols:
            if pd.notna(row[col]):
                set_clauses.append(f"{col} = :{col}")
                params[col] = row[col]

        if set_clauses:
            params['film_id'] = int(row['film_id'])
            sql_update = text(f"""
                UPDATE film
                SET {', '.join(set_clauses)}
                WHERE film_id = :film_id;
            """)
            
            conn.execute(sql_update, params)

In [67]:
# Example query
run_query("SELECT * FROM film WHERE title = 'Inception';")

film_id      title release_date  vote_average  vote_count  runtime  \
0     1429  Inception   2010-07-15         8.364       34495      148   

       revenue       budget  
0  825532764.0  160000000.0

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Let's now add 100,000 random films from the movie dataset.
</div>

In [68]:
df_sample = films.sample(n=10000, random_state=42)
df_sample = df_sample[['title', 'release_date', 'vote_average', 'vote_count', 'runtime', 'revenue', 'budget']]
df_sample

title release_date  \
36225                                            Adalen 31   1969-05-01   
474786             Queers, Christians and Canadian Justice   2019-10-18   
964430                                            Fasching   1939-09-13   
7700                                         Kajillionaire   2020-09-25   
1022218  Calexico: World Drifts In (Live at The Barbica...   2004-10-12   
...                                                    ...          ...   
776507                                  Давай без распятия   2021-12-21   
305993         Urban Legacy: The Story Behind Urban Legend   2018-11-20   
170293                           Recollections of Pavlovsk   1984-06-01   
86015                                       One Man's Loss   2012-01-01   
349633                              Hitler's Secret Weapon   2010-06-21   

         vote_average  vote_count  runtime   revenue  budget  
36225           6.800          33    114.0       NaN     NaN  
474786            NaN           0     49.0       NaN     NaN  
964430            NaN           0      NaN       NaN     NaN  
7700            6.286         383    105.0  829900.0     NaN  
1022218           NaN           0    158.0       NaN     NaN  
...               ...         ...      ...       ...     ...  
776507            NaN           0      NaN       NaN     NaN  
305993          7.000           1    147.0       NaN     NaN  
170293          5.700           3     39.0       NaN     NaN  
86015           7.400           8      8.0       NaN     NaN  
349633          7.000           1     46.0       NaN     NaN  

[10000 rows x 7 columns]

In [69]:
df_sample = df_sample.replace({np.nan: None})

In [70]:
sql_insert = text(f"""
     INSERT INTO films.film (title, release_date, vote_average, vote_count, runtime, revenue, budget)
     VALUES (:title, :release_date, :vote_average, :vote_count, :runtime, :revenue, :budget)
     ON CONFLICT ON CONSTRAINT uq_film_title_release DO NOTHING;""")

with engine.begin() as conn:
     conn.execute(sql_insert, df_sample.to_dict('records'))

# Queries (part 2)

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Get the films released in 2020 or later.
</div>

In [71]:
run_query("""
SELECT title, release_date
FROM film
WHERE release_date >= '2020-01-01'
LIMIT 10;
""")

title release_date
0    GamerGate: The Untold Story   2021-01-01
1                    Dying Alone   2024-10-10
2                       Fat Rick   2024-09-06
3                       Vultures   2021-02-01
4                           HAND   2021-04-12
5                           Pink   2021-04-23
6        Dr. Hot's Family Clinic   2024-05-22
7                      Jai Bheem   2022-04-13
8               The Doppelgänger   2023-06-15
9  La Determinación del Devenir   2020-09-08

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Get the films with vote_average below 6 but that still have been nominated for Oscar.
</div>

In [167]:
run_query("""
SELECT DISTINCT f.title, f.vote_average, n.is_winner
FROM film f
JOIN nomination n USING(film_id)
WHERE vote_average < 6 AND vote_count > 1000;
""")

title  vote_average  is_winner
0                102 Dalmatians         5.456      False
1          Fifty Shades of Grey         5.882      False
2                Into the Woods         5.750      False
3                 Hail, Caesar!         5.914      False
4                      Poseidon         5.841      False
5              Superman Returns         5.738      False
6         Babe: Pig in the City         5.553      False
7              The Midnight Sky         5.766      False
8                        Norbit         5.584      False
9                 Mirror Mirror         5.918      False
10          The Nutty Professor         5.625       True
11  Snow White and the Huntsman         5.982      False
12             The Black Dahlia         5.621      False
13               Batman Forever         5.409      False
14                   Hollow Man         5.918      False
15                Suicide Squad         5.909       True
16                       Junior         5.167      False
17                  The Wolfman         5.842       True

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Average budget of nominated films with Brad Pitt.
</div>

In [77]:
run_query("""
WITH brad_movie AS (
    SELECT n.film_id
    FROM nomination n
    JOIN nomination_person np ON n.nom_id = np.nomination_id
    JOIN person p USING(person_id)
    WHERE p.name = 'Brad Pitt'
)
SELECT AVG(f.budget)
FROM film f
JOIN brad_movie USING(film_id);
""")

avg
0  59600000.0

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Find the film that won an Oscar with the lowest budget.
</div>

In [102]:
run_query("""
    SELECT f.title, f.budget
    FROM film f
    JOIN nomination n ON f.film_id = n.film_id
    JOIN award a ON n.award_id = a.award_id
    WHERE n.is_winner AND f.budget != 'NaN'
    ORDER BY  f.budget ASC
    LIMIT 1;
;
"""
)

title  budget
0  Kiss of the Spider Woman    11.0

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Get top 10 movies (we add vote_count constraint for relevance).
</div>

In [128]:
run_query("""
with film_win as ( select film_id, count(*) as nb_win
from nomination where is_winner=true group by film_id)
SELECT title, vote_average, nb_win
FROM film F JOIN film_win fw USING (film_id)
WHERE vote_average IS NOT NULL AND vote_count > 10000
ORDER BY vote_average DESC
LIMIT 10;
""")

title  vote_average  nb_win
0                                  The Godfather         8.707       3
1                          The Godfather Part II         8.591       6
2                               Schindler's List         8.573       7
3                                  Spirited Away         8.539       1
4                                       Parasite         8.515       4
5                                The Dark Knight         8.512       2
6                                   Pulp Fiction         8.488       1
7                                   Forrest Gump         8.477       6
8  The Lord of the Rings: The Return of the King         8.474      11
9                              Life Is Beautiful         8.455       3

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Find the average note for person who worked/acted in more than 10 movies.
</div>

In [118]:
run_query("""
SELECT p.name AS Actor_Name, AVG(f.vote_average) AS average_film_rating
FROM Person p
JOIN Nomination_Person np ON p.person_id = np.person_id
JOIN Nomination n ON np.nomination_id = n.nom_id
JOIN Award a ON a.award_id = n.award_id
JOIN Film f ON n.film_id = f.film_id
WHERE f.vote_average IS NOT NULL
GROUP BY p.name
HAVING COUNT(DISTINCT f.film_id) > 10
ORDER BY average_film_rating DESC
LIMIT 10; 
""")

actor_name  average_film_rating
0   Michael Semanick             7.814818
1       Billy Wilder             7.727278
2  Christopher Boyes             7.552200
3  Alexandre Desplat             7.508091
4        Hans Zimmer             7.498818
5        Andy Nelson             7.494680
6      Roger Deakins             7.490429
7   Steven Spielberg             7.488818
8         Randy Thom             7.483600
9        Joe Letteri             7.444455

<div style="border: 2px solid #4CAF50; padding: 10px; border-radius: 8px; background-color:#f0fff0">
Find the director who won the most oscar</div>

In [142]:
run_query("""
SELECT p.name AS Director_Name,COUNT(n.nom_id) AS nb_best_director
FROM Person p
JOIN Nomination_Person np ON p.person_id = np.person_id
JOIN Nomination n ON np.nomination_id = n.nom_id
JOIN Award a ON n.award_id = a.award_id
WHERE n.is_winner = TRUE AND a.award_name LIKE 'Directing'
GROUP BY p.name ORDER BY nb_best_director DESC 
LIMIT 1;
""")

director_name  nb_best_director
0     John Ford                 4